clang/test/Sema/flexible-array-in-union.c

```diff
@@ -1,13 +1,58 @@
-// RUN: %clang_cc1 %s -verify=c -fsyntax-only
+// RUN: %clang_cc1 %s -verify -fsyntax-only
 // RUN: %clang_cc1 %s -verify -fsyntax-only -x c++
-// RUN: %clang_cc1 %s -verify -fsyntax-only -fms-compatibility
 // RUN: %clang_cc1 %s -verify -fsyntax-only -fms-compatibility -x c++
+// RUN: %clang_cc1 %s -verify=gnu -fsyntax-only -Wgnu-flexible-array-union-member -Wgnu-empty-struct
+// RUN: %clang_cc1 %s -verify=microsoft -fsyntax-only -fms-compatibility -Wmicrosoft
 
 // The test checks that an attempt to initialize union with flexible array
 // member with an initializer list doesn't crash clang.
 
 
-union { char x[]; } r = {0}; // c-error {{flexible array member 'x' in a union is not allowed}}
+union { char x[]; } r = {0}; /* gnu-warning {{flexible array member 'x' in a union is a GNU extension}}
+                                microsoft-warning {{flexible array member 'x' in a union is a Microsoft extension}}
+                              */
```

Comment:
If this is no longer an error, and the primary use case of this is for unions that are members of other structures, what are the semantics for initialization of flexible array members in unions in other structures? Is this covered sufficiently by existing tests?

Reply:
Clang's initialization for flexible arrays works for all the cases I've checked. Part of Linux's migration to C99 flex arrays uncovered a `__builtin_object_size()` bug that has was fixed in Clang 17 via commit 804af933f7310f78d91e13ad8a13b64b00249614 which includes several initialization tests beyond the ones that are touched in this PR. As a result I have confidence in this having appropriate test coverage, though if more is needed, it should be trivial to add.

Reply:
I do not see any obvious cases where the code does the wrong thing, sure, I agree with you on that, but as far as I can find, we do not have *any* test for "initialization of flexible array member is not allowed" being correctly emitted for nested unions containing flexible array members, either currently or after this PR. `flexible-array-init.c` contains tests for this for nested structs containing flexible array members, but not for unions. Previously, these would have implicitly been handled by flexible array members being disallowed in unions in the first place, at least for C, but as you are enabling them now, that no longer applies.

Reply:
Ah-ha! I think I see what you mean now. I will work on constructing some negative tests cases.

Reply:
I've added a bunch more tests now.